pkg/storage/cacher_test.go

```diff
@@ -344,3 +345,53 @@ func TestFiltering(t *testing.T) {
 	verifyWatchEvent(t, watcher, watch.Modified, podFooPrime)
 	verifyWatchEvent(t, watcher, watch.Deleted, podFooPrime)
 }
+
+func TestStartingResourceVersion(t *testing.T) {
+	server, etcdStorage := newEtcdTestStorage(t, testapi.Default.Codec(), etcdtest.PathPrefix())
+	defer server.Terminate(t)
+	cacher := newTestCacher(etcdStorage)
+	defer cacher.Stop()
+
+	// add 1 object
+	podFoo := makeTestPod("foo")
+	fooCreated := updatePod(t, etcdStorage, podFoo, nil)
+
+	// Set up Watch starting at fooCreated.ResourceVersion + 10
+	rv, err := storage.ParseWatchResourceVersion(fooCreated.ResourceVersion)
+	if err != nil {
+		t.Fatalf("Unexpected error: %v", err)
+	}
+	rv += 10
+	startVersion := strconv.Itoa(int(rv))
+
+	watcher, err := cacher.Watch(context.TODO(), "pods/ns/foo", startVersion, storage.Everything)
+	if err != nil {
+		t.Fatalf("Unexpected error: %v", err)
+	}
+	defer watcher.Stop()
+
+	lastFoo := fooCreated
+	for i := 0; i < 10; i++ {
+		podFooForUpdate := makeTestPod("foo")
+		podFooForUpdate.Labels = map[string]string{"foo": strconv.Itoa(i)}
+		lastFoo = updatePod(t, etcdStorage, podFooForUpdate, lastFoo)
+	}
+
+	podFooForUpdate := makeTestPod("foo")
+	podFooForUpdate.Labels = map[string]string{"foo": "bar"}
+	updatedFoo := updatePod(t, etcdStorage, podFooForUpdate, lastFoo)
+
+	select {
+	case e := <-watcher.ResultChan():
+		o := e.Object.(*api.Pod)
+		// event should have rv + 1, since we're starting the watch at rv
+		if e, a := strconv.Itoa(int(rv+1)), o.ResourceVersion; e != a {
```

Comment:
I remember that at some point we have unit test flakes because etcd itself is internally increasing ResourceVersion (without any external action on it). This is still true during startup (RV from the first external action is usually 3).

@xiang90 @hongchaodeng - is that safe to assume that of we have 12 creates/updates to etcd, these will have consecutive RV numbers?

@ncdc - to be on the safe side, I would just check whether the RV is >rv, not exactly equal rv+1 (and then just ignore checking equality of objects)